cmd/metalctl/app/move.go

```diff
@@ -0,0 +1,226 @@
+	"github.com/spf13/cobra"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	sourceKubeconfig string
+	targetKubeconfig string
+	crdsOnly         bool
+	crsOnly          bool
+	namespace        string
+	dryRun           bool
+	verbose          bool
+)
+
+func NewMoveCommand() *cobra.Command {
+	move := &cobra.Command{
+		Use:   "move",
+		Short: "Move metal-operator CRDs and CRs from one cluster to another",
+		RunE:  runMove,
+	}
+	move.Flags().StringVar(&sourceKubeconfig, "source-kubeconfig", "", "Kubeconfig pointing to the source cluster")
+	move.Flags().StringVar(&targetKubeconfig, "target-kubeconfig", "", "Kubeconfig pointing to the target cluster")
+	move.Flags().BoolVar(&crdsOnly, "crds-only", false, "migrate only the CRDs without CRs")
+	move.Flags().BoolVar(&crsOnly, "crs-only", false, "migrate only the CRs without CRDs")
+	move.Flags().StringVar(&namespace, "namespace", "", "namespace to filter CRDs and CRs to migrate. Defaults to all namespaces if not specified")
+	move.Flags().BoolVar(&dryRun, "dry-run", false, "show what would be moved without executing the migration")
+	move.Flags().BoolVar(&verbose, "verbose", false, "enable verbose logging for detailed output during migration")
+	move.MarkFlagRequired("source-kubeconfig")
+	move.MarkFlagRequired("target-kubeconfig")
+
+	if verbose {
+		slog.SetLogLoggerLevel(slog.LevelDebug)
+	}
+	return move
+}
+
+type Clients struct {
+	source client.Client
+	target client.Client
+}
+
+func makeClient(kubeconfig string) (client.Client, error) {
+	cfg, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load cluster kubeconfig: %w", err)
+	}
+	return client.New(cfg, client.Options{Scheme: scheme})
+}
+
+func makeClients() (Clients, error) {
+	var clients Clients
+	var err error
+
+	clients.source, err = makeClient(sourceKubeconfig)
+	if err != nil {
+		return clients, fmt.Errorf("failed to construct source cluster client: %w", err)
+	}
+	clients.target, err = makeClient(targetKubeconfig)
+	if err != nil {
+		return clients, fmt.Errorf("failed to construct target cluster client: %w", err)
+	}
+	return clients, nil
+}
+
+// getMetalObjects returns CRDs and CRs from metal group. CRDs in a returned list are before theirs CRs.
+func getMetalObjects(ctx context.Context, cl client.Client) ([]client.Object, error) {
+	crds := &apiextensionsv1.CustomResourceDefinitionList{}
+	if err := cl.List(ctx, crds); err != nil {
+		return nil, fmt.Errorf("couldn't list CRDs: %w", err)
+	}
+
+	metalObjects := make([]client.Object, 0)
```

Comment:
Maybe it is worth to store CRDs and CRs in 2 different slices? That way, type checks/branching while iterating them can be avoided later.